x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/Wildcard.java

```diff
@@ -0,0 +1,142 @@
+import org.elasticsearch.xpack.ql.expression.gen.pipeline.Pipe;
+import org.elasticsearch.xpack.ql.expression.gen.script.ScriptTemplate;
+import org.elasticsearch.xpack.ql.expression.predicate.logical.Or;
+import org.elasticsearch.xpack.ql.expression.predicate.regex.Like;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+
+import java.util.Collections;
+import java.util.List;
+
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isStringAndExact;
+
+/**
+ * EQL wildcard function. Matches the form:
+ *     wildcard(field, "*wildcard*pattern*", "*wildcard*pattern*")
+ */
+
+public class Wildcard extends ScalarFunction {
+
+    private final Expression field;
+    private final List<Expression> patterns;
+
+    private static List<Expression> getArguments(Expression src, List<Expression> patterns) {
+        List<Expression> arguments = new java.util.ArrayList<>(Collections.singletonList(src));
+        arguments.addAll(patterns);
+        return arguments;
+    }
+
+    public Wildcard(Source source, Expression field, List<Expression> patterns) {
+        super(source, getArguments(field, patterns));
+        this.field = field;
+        this.patterns = patterns;
+    }
+
+    public Expression asLikes() {
+        Expression result = null;
+
+        for (Expression pattern: patterns) {
+            String wcString = pattern.fold().toString();
+            Like like = new Like(source(), field, StringUtils.toLikePattern(wcString));
+            result = result == null ? like : new Or(source(), result, like);
+        }
+
+        return result;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (!childrenResolved()) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution lastResolution = isStringAndExact(field, sourceText(), ParamOrdinal.FIRST);
+        if (lastResolution.unresolved()) {
+            return lastResolution;
+        }
+
+        for (Expression p: patterns) {
+            lastResolution = isStringAndExact(p, sourceText(), ParamOrdinal.DEFAULT);
+            if (lastResolution.unresolved()) {
+                break;
+            }
+
+            if (p.foldable() == false) {
+                return new TypeResolution(format(null, "wildcard against variables are not (currently) supported; offender [{}] in [{}]",
+                    Expressions.name(p),
+                    sourceText()));
+            }
+        }
+
+        return lastResolution;
+    }
+
+    @Override
+    protected Pipe makePipe() {
+        Expression asLikes = asLikes();
+        if (asLikes instanceof Like) {
+            return ((Like) asLikes).makePipe();
+        } else {
+            return ((Or) asLikes).makePipe();
+        }
```

Comment:
```suggestion
    return asLike().makePipe();
```

Reply:
This doesn't seem to be needed and in fact the implementation should be `throw new EqlIllegalArgumentException("should not call this method")`

Reply:
awesome. that was my hope.
I still need to keep `fold()` though for the optimizer, right?